app/controllers/checkouts_controller.rb

```diff
@@ -0,0 +1,78 @@
+class CheckoutsController < ApplicationController
+  before_action :set_variables, only: [:checkout_current_card, :checkout_new_card]
+
+
+  def new
+    @product = Product.find(params[:id])
+    @cards = Card.where(user: current_user)
+  end
+
+
+  def checkout_current_card
+    @charge = Stripe::Charge.create(
+      :customer => params[:customer_id],
+      :amount => @amount,
+      :currency => 'eur'
+    )
+
+    record_purchase
+
+  rescue Stripe::CardError => e
+    flash[:error] = e.message
+    redirect_to root_path
+  end
+
+
+  def checkout_new_card
+    # create the customer account on stripe
+    customer = Stripe::Customer.create(
+      :email => params[:stripeEmail],
+      :source  => params[:stripeToken]
+    )
+
+    @charge = Stripe::Charge.create(
+      :customer    => customer.id,
+      :amount      => @amount,
+      :description => 'Marketplace customer',
+      :currency    => 'eur'
+    )
+
+    # store the user and the customer_id from stripe so we can charge the card automatically
+    card = Card.create(user: current_user, stripe_customer_id: customer.id, stripe_card_id: @charge[:source][:id], card_brand: @charge[:source][:brand], card_last_4: @charge[:source][:last4])
+
+    record_purchase
+
+  rescue Stripe::CardError => e
+    flash[:error] = e.message
+    redirect_to root_path
+  end
+
+
+
+  private
+  def set_variables
+    @product = Product.find(params[:id])
+    @shop = @product.shop
+    @amount = @product.price.to_i * 100
+  end
+
+  def record_purchase
+    # create record of the purcase in the database
+    purchase = Purchase.create(user: current_user, product: @product, purchasing_price: @product.price, purchasing_quantity: 1, stripe_charge_id: @charge.id)
+
+    # balance transaction information for a particular charge
+    balance_transaction = Stripe::BalanceTransaction.retrieve(@charge[:balance_transaction])
+    balance_after_stripe_fees = balance_transaction[:net].to_i
+    balance_after_marketplace_percentage = balance_after_stripe_fees - marketplace_percentage(balance_after_stripe_fees)
+    new_shop_balance = @shop.balance.to_i + balance_after_marketplace_percentage
+    @shop.update(balance: new_shop_balance)
```

Comment:
Yes, but we can't merge this branch then which will have this code tho. Each branch must be complete in itself that once merged it could be deployed if necessary. Each branch should not be dependent on each other. If this is only going to be fixed on the withdrawal feature, better to do the withdrawal in this branch too.